File: [A]tree/[A]tree-dfs-postorder/543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        res = 0

        def dfs(node):
            nonlocal res
            if not node:
                return 0
            left_path = dfs(node.left)
            right_path = dfs(node.right)
            node_as_peak = left_path + right_path
            res = max(res, node_as_peak)
            return max(left_path + 1, right_path + 1)

        dfs(root)
        return res
        
# time O(n), due to traverse
# space O(n), due to tree height for skewed tree
# using tree and dfs (postorder, recursive)

File: [A]tree/[A]tree-dfs-postorder/543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

    int res;

    public int diameterOfBinaryTree(TreeNode root) {
        res = 0;
        dfs(root);
        return res;
    }

    public int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int leftL = dfs(node.left);
        int rightL = dfs(node.right);
        int nodeAsPeakL = leftL + rightL;
        res = Math.max(res, nodeAsPeakL);
        return Math.max(leftL + 1, rightL + 1);
    }
}

// time O(n), due to traverse
// space O(n), due to tree height for skewed tree
// using tree and dfs (postorder, recursive)